pkg/modprovider/views.go

```diff
@@ -0,0 +1,242 @@
+// Copyright 2016-2025, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package modprovider
+
+import (
+	"errors"
+
+	"github.com/pulumi/pulumi-terraform-module/pkg/tfsandbox"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+	pulumirpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
+	"google.golang.org/protobuf/types/known/structpb"
+)
+
+func viewStepsPlan(
+	packageName packageName,
+	plan *tfsandbox.Plan,
+) []*pulumirpc.ViewStep {
+	return viewStepsGeneric(packageName, plan, nil)
+}
+
+func viewStepsAfterApply(
+	packageName packageName,
+	plan *tfsandbox.Plan,
+	appliedState *tfsandbox.State,
+) []*pulumirpc.ViewStep {
+	return viewStepsGeneric(packageName, plan, appliedState)
+}
+
+func viewStepsAfterRefresh(
+	packageName packageName,
+	plan *tfsandbox.Plan,
+	refreshedState *tfsandbox.State,
+) []*pulumirpc.ViewStep {
+	return viewStepsGeneric(packageName, plan, refreshedState)
+}
+
+func viewStepsGeneric(
+	packageName packageName,
+	plan *tfsandbox.Plan,
+	finalState *tfsandbox.State,
+) []*pulumirpc.ViewStep {
+	var steps []*pulumirpc.ViewStep
+	priorState, hasPriorState := plan.PriorState()
+	hasFinalState := finalState != nil
+
+	plan.VisitResources(func(rplan *tfsandbox.ResourcePlan) {
+		// TODO sometimes addresses change but identity remains the same.
+		addr := rplan.Address()
+
+		var priorRState, finalRState *tfsandbox.ResourceState
+
+		if hasPriorState {
+			s, ok := priorState.FindResource(addr)
+			if ok {
+				priorRState = s
+			}
+		}
+
+		if hasFinalState {
+			s, ok := finalState.FindResource(addr)
+			if ok {
+				finalRState = s
+			}
+		}
+
+		step := viewStepForResource(packageName, rplan, priorRState, finalRState)
+		steps = append(steps, step)
+	})
+	return steps
+}
+
+func viewStepOp(changeKind tfsandbox.ChangeKind) pulumirpc.ViewStep_Op {
+	switch changeKind {
+	case tfsandbox.NoOp:
+		return pulumirpc.ViewStep_SAME
+	case tfsandbox.Update:
+		return pulumirpc.ViewStep_UPDATE
+
+	// resource is replaced by Creating a new copy then deleting the old copy
+	case tfsandbox.Replace:
+		return pulumirpc.ViewStep_DELETE_REPLACED // TODO is this right?
```

Comment:
@justinvp would you know what should we be pumping into views for the two kinds of replace? Should we be sending the views twice, or even three times (one for preview, one for create, one for delete for example).

Reply:
Publish three view steps in sequence to represent create before delete:

1. `CREATE_REPLACEMENT`
2. `REPLACE`
3. `DELETE_REPLACED`

And for delete before replace:

1. `DELETE_REPLACED`
2. `REPLACE`
3. `CREATE_REPLACEMENT`

Emulating what the step generator would do: https://github.com/pulumi/pulumi/blob/c5ad4f800a4ce907e4bd5e4e9df73d603719e855/pkg/resource/deploy/step_generator.go#L1543-L1652